### PR TITLE
Refactor major components to streamline logic

### PR DESCRIPTION
**Pull request by Gozala**
## Problem Statement

As I worked on #1568 I had hard time figuring out what was happening where, in fact I [left a TODO there](https://github.com/ipfs-shipyard/ipfs-webui/pull/1569#discussion_r467883063) because I was unable to figure out what `make` function in the `bundles/files/utils` was trying to do:

https://github.com/ipfs-shipyard/ipfs-webui/blob/16e15a25ab999d4912ea05ebb81e9bad67457770/src/bundles/files/utils.js#L12-L78

Then as I was trying to find what listenes to [`FILES_WRITE_UPDATED`](https://github.com/ipfs-shipyard/ipfs-webui/blob/16e15a25ab999d4912ea05ebb81e9bad67457770/src/bundles/files/actions.js#L303) actions, I was able to find nothing. Later I discovered that `make` function does some trickery where it dispatches `FILES_X_STARTED`, `FILES_X_FINISHED` actions and also does some extra stuff on finish. This all makes really difficult to know what happens where.

### Solution

This pull request is attempt to streamline everything, so that looking at the specific `doX` function tells you all you need to know without having to look around other places to get a sense what happens after finish etc... In order to accomplish that it switches from `FILES_X_STARTED` / `FILES_X_FINISHED` style action dispatch to making `FILES_X` actions dispatch events with `job` field that that updates it's state (as in state machine) from `Idle` -> `Active` -> (`Done` | `Failed`). 

This also gets rid of all the post success followups:

https://github.com/ipfs-shipyard/ipfs-webui/blob/16e15a25ab999d4912ea05ebb81e9bad67457770/src/bundles/files/utils.js#L42-L60

And post failure followups:

https://github.com/ipfs-shipyard/ipfs-webui/blob/16e15a25ab999d4912ea05ebb81e9bad67457770/src/bundles/files/utils.js#L61-L76

By using `try { ... } finally { ... }` in in the `doX` functions, so it's plain and clear:

https://github.com/ipfs-shipyard/ipfs-webui/blob/f235c331c80068a023359efd50cbef48a6008d9b/src/bundles/files/actions.js#L324-L338

Pull request also introduces idea of non atomic tasks that (start, make progress, finish or fail) so that `doX` tasks that report progress don't need to do custom workarounds:

https://github.com/ipfs-shipyard/ipfs-webui/blob/16e15a25ab999d4912ea05ebb81e9bad67457770/src/bundles/files/actions.js#L298-L310

Instead spawned tasks (that follow above described state machine) can yield, to produce update states:

https://github.com/ipfs-shipyard/ipfs-webui/blob/f235c331c80068a023359efd50cbef48a6008d9b/src/bundles/files/actions.js#L243-L251

And it adds typing for all of this so any inconsistencies could caught early on. In the process I also fixed bunch that existed.

### CODE REVIEW

**Comment by Gozala** [comment]
> Small ask: as noted during our chat today, we need to understand if/how this change impacts the way WebUI's opt-in Countly metrics are gathered (not a blocker, but we should know the impact before merging this).

This was a good call @lidel, indeed analytics seems to use mix of regexps and strings to hook into these actions, which is also a reason I have missed those:

https://github.com/ipfs-shipyard/ipfs-webui/blob/master/src/bundles/analytics.js#L6-L20

I will look into updating that module to take into account the changes made here.

**Review by lidel — COMMENTED** [review]
@Gozala I am mildly in favor, but afraid I am not aware of all ramifications of this refactor.
Mind scheduling a call with me and @olizilla this or next week to go over changes and unblock this?

**Review by rafaelramalho19 — CHANGES_REQUESTED** [review]

File: src/bundles/ipfs-provider.js
Comment (rafaelramalho19) [inline src/bundles/ipfs-provider.js:64]: I would invert this into early exit

```
if (!result.ok) {
 return { ...state, ready: false, failed: true }
}
// ...
```

**Comment by lidel** [comment]
@Gozala mind rebasing this PR?
We really want to merge this before https://github.com/ipfs-shipyard/ipfs-webui/pull/1615 (if possible)

File: src/bundles/ipfs-provider.js
Comment (Gozala) [reply src/bundles/ipfs-provider.js]: Turns out what redux-bundle passes into `doX` style things isn't a store but rather it's own context thing that contains `store` (see https://reduxbundler.com/api-reference/bundle#bundle-doxA) I think I originally was calling it a store as well, but then it got confusing because in some cases it is store and in others it isn't.

File: src/bundles/files/utils.js
Comment (Gozala) [reply src/bundles/files/utils.js:148]: > What's the point of creating a function that just serves as a way to invoke two variables?

That's fare, but I would go step further and ask what's the point of having redux-bundler with all the actions creators function that returns the function that is passed the context so that it can just dispatch the action ?

```js
const actions = {
  ...,
  doFilesDismissErrors: () => async (context) => context.store.dispatch({
     type: ACTIONS.DISMISS_ERRORS
   }),
   ...
}
```
I just find not having to create intermediate function manually a bit more clear and it reduces boilerplate. 

```js
{
  ...,
  doFilesDismissErrors: () => send({ type: ACTIONS.DISMISS_ERRORS }),
  ...
}
```

That said I do find it style that redux-bundle forces quite a bit annoying.

> Also, I don't see where this function is being used 🤔

It is used quite a bit in the filse/actions.js

File: src/bundles/ipfs-provider.js
Comment (Gozala) [reply src/bundles/ipfs-provider.js:1]: Indeed

File: src/bundles/ipfs-provider.js
Comment (Gozala) [reply src/bundles/ipfs-provider.js:64]: Why ?

File: src/bundles/ipfs-provider.js
Comment (Gozala) [reply src/bundles/ipfs-provider.js:179]: Because setting could be a `null` (e.g. if try above failed), while `JSON.parse` has type of `(input:string) -> JSON`. Passing `string|null` would be rejected by a type checker. `setting || ''` on the other hand has `string` type.

I would say that error in this scenario is an expected behavior.

File: src/bundles/ipfs-provider.js
Comment (Gozala) [reply src/bundles/ipfs-provider.js:182]: Well return type of this function is `string | null` so `null` is fine. It maybe reasonable to change type to `string` and check for `''`, but it would be bit trickier to catch all the places where `''` case isn't considered because type checker won't really help, unlike spotting cases where `string | null` is treated as string.

File: tsconfig.json
Comment (Gozala) [reply tsconfig.json:60]: 😭 I can't seem to figure out why my editor seems to do that with json files. I'll try to figure that out

**Comment by Gozala** [comment]
> @Gozala mind rebasing this PR?
> We really want to merge this before #1615 (if possible)

Just pushed the version that merges changes from master. It end up been a bit more work than I have anticipated & I had not had a chance to do a testing with merged version yet.

**Review by lidel — APPROVED** [review]
LGTM, no regressions as far I can tell, and should make progress reporting easier to reason about.

@Gozala what would be the next step for having visual feedback during file import? Any PRs/issues in other repos that I could help landing/unblocking?

@rafaelramalho19 please block some time this week to do another pass at this

File: @types/redux-bundler/index.d.ts
Comment (Gozala) [reply @types/redux-bundler/index.d.ts:15]: I don't believe it is worth it honestly. Because it is not really possible to add proper typings to the redux-bundler. What I do here is just provide some subset which also requires very particular use of redux-bundler (splitting actions and selectors, which seems to go against it's design goals), and even then have to manually annotate.

2. Even if we add types there, which I doubt would be accepted due to mentioned limitations, I don't think the effort to do that is going to be worth it.

I think if we were to invest more time, it would be probably best to consider using something that is more type friendly / ready than redux-bundler is.

File: src/bundles/files/index.js
Comment (rafaelramalho19) [inline src/bundles/files/index.js:159]: ⚠️ There's a typo, it should be called `pullPending`

File: src/bundles/analytics.js
Comment (rafaelramalho19) [inline src/bundles/analytics.js:92]: ⚠️ No need to comment these 2 lines, you can just remove them

**Review by rafaelramalho19 — CHANGES_REQUESTED** [review]
After the 2 small nits it's good to merge.

Good job 👍

**Pull request merged**
